x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportStopTransformAction.java

```diff
@@ -242,6 +242,17 @@ protected void taskOperation(
         }
 
         if (ids.contains(transformTask.getTransformId())) {
+            if (request.isForce()) {
+                try {
+                    // If force==true, we skip all the additional steps and only deregister and complete persistent task via shutdown call.
+                    // This way we ensure that the persistent task is removed ASAP (as opposed to being removed in one of the listeners).
+                    transformTask.shutdown();
```

Comment:
I'm not an expert in how this all fits together, so if there's a reason this idea doesn't work then don't implement it.

It seems that `transformTask.shutdown()` will stop the persistent task and stop the transform being scheduled again. But it _won't_ stop the indexer object from continuing to work. So possibly immediately after this line we could have this:

```
    if (transformTask.getIndexer() != null) {
        transformTask.getIndexer().abort();
    }
```

Reply:
The idea of `onCancelled` is that it gets called if the persistent task is removed from another node. For example, suppose the master node removed the persistent task, then the persistent tasks node service running on the transform node would notice that and call `onCancelled` on the transform node.

`shutdown` calls `markAsCompleted`, which is the transform node itself saying that the persistent task has ended gracefully. Because this is the node where the persistent task is running saying that the persistent task is complete I doubt `onCancelled` will then automatically get called by the framework, but I might be wrong. I just added the `cloud-deploy` label to this PR. What you can do to check is to start a Cloud cluster in the CFT region using the custom Docker image, enable debug logging for transforms in that cluster and then experiment with starting and force-stopping transforms. From the debug in the `onCancelled` method you'll be able to see whether it now gets called once or twice. Looking at the debug for a real cluster running this change may also reveal something else that doesn't look quite right following this change.